### DISCUSSION TITLE
did the neovim project choose the right policy with regard to GUI development?

### DISCUSSION THREAD

**Discussion by JamesWidman** [neovim/neovim#25002] [General]
Hi all,

I've been using some GUI version of some version of vim since 1999. And I've been using a vim-or-neovim GUI on the mac for about 20 years.

Since 2003, my main motive for using a GUI vim on mac OS is that it allowed the use of any combination of the four modifier keys found on mac keyboards (command, option, control, & shift).

With neovim, [i don't know if that's possible anymore](https://github.com/neovim/neovim/discussions/24998), and I can't help but think that this might be related to the difference between Neovim and Bram's Vim in terms of their approaches to GUIs.

Like, if we look at [the mac section on vim's download page](https://www.vim.org/download.php#mac), we find one GUI.

*One*.

And last I checked, it supported the use of modifier keys.

In contrast, if we look at [the list of GUIs for Neovim](https://github.com/neovim/neovim/wiki/Related-projects#gui), we find a number of GUI projects. Of those projects that are (1) potentially for the mac* and (2) currently maintained, I have not yet found one that supports the use of all modifier keys.

So this leads me to ask: did neovim's policy on GUIs actually produce a user experience that is at least as good as our experience with Bram's vim? Because to me it seems that the combination of:

1) exposing a public API for GUIs, and
2) not producing an official GUI as part of the neovim project (or at least endorsing one GUI as the "official" one) 

...has just splintered efforts. Instead of organizing GUI programmers to work on one project for all neovim users on each operating system, these programmers have been encouraged to spread themselves out over many different projects, like butter scraped over too much bread. And consequently, some of these vim GUIs just *don't work*, or only partially work. (Or: they work, but only for an old version of neovim.)

Given the loss of functionality and the proliferation of unmaintained GUI projects, I think it's fair to ask: is the GUI policy of the neovim project actually working? 

---
[*] side note: I'm not sure how many of the GUIs mentioned on the [GUI list](https://github.com/neovim/neovim/wiki/Related-projects#gui) are supposed to work on macOS.  For example, there's a "gtk4" version, and gtk has been ported to macos, but I don't know if this program has been tested on macOS.

side-side note: We can agree that "gtk4" is not a *platform*, right? Like, as a user, I don't necessarily care what API you use to open a window and render glyphs; I only care whether it will run on the same *operating system and hardware* that I'm currently running. And if an app is ostensibly "cross-platform" because the GUI developer used a cross-platform API, that doesn't tell me whether they *tested* it on the OS I use. So mentioning an *implementation detail* like "gtk" in the "platform" column of this table is not helpful.  (Though, if there were a separate table for people who want to contribute to the development of neovim GUIs, that might be a good place to mention that it was implemented with gtk or Electron or whatever.)

**Comment by clason** [comment]
> I think it's fair to ask: is the GUI policy of the neovim project actually working?

Yes. There are many active projects addressing different audiences, and they are not held back by the core maintainers' finite resources.

**Reply by JamesWidman** [reply]
I respect that the core maintainers' resources are finite.

It seems like it should be possible to avoid assigning core maintainers to any GUI project while also somehow  encouraging a consolidation of efforts on GUIs so that there can be a working GUI on macOS (which we don't seem to have at this time, which is a regression relative to Vim).

**Reply by clason** [reply]
> encouraging a consolidation of efforts on GUIs

Why? That is a fallacy (the same as the constant cry for "you should merge efforts with Vim!"). People work on what they want to work on, and you can't force them to work on the GUI _you_ would like to use. Many GUIs are created because the dev had a _different_ vision from previous efforts.

And Neovide works just fine on macOS. (As does VimR, for that matter.)

**Reply by JamesWidman** [reply]
Maybe "working" sounds like a loaded way to put it, but like I said: I have 20 years' worth of keybindings. I don't like the idea of giving up on that!

**Reply by clason** [reply]
So you have a very narrow definition of "working". Pick a GUI and (help them) implement what you need there. (Spoiler: input handling is the trickiest part of an editor.)

And if Vim fits your use case better, by all means keep using it! Neovim does not aim to replace Vim but to extend its philosophy in new directions.

**Reply by JamesWidman** [reply]
1. vimr seems to be unmaintained since January. This might not be too bad, but each version of vimr is tied to a particular version of neovim.

2. I was using neovide until today. A month ago, they broke `<command-shift-` bindings, [which is a known issue in version 0.11](https://github.com/neovide/neovide/releases) (see Breaking Changes).  As a workaround  i reinstalled the older version.  But yesterday, neovide is just... not launching anymore.  I will try to debug it, but I suspect that they're just not testing on macOS.

**Reply by clason** [reply]
> But yesterday, neovide is just... not launching anymore

Works for me. Did you open an issue and offer to help with testing and fixing?

**Reply by JamesWidman** [reply]
> And if Vim fits your use case better, by all means keep using it! Neovim does not aim to replace Vim but to extend its philosophy in new directions.

The thing is, I've been using a neovim GUI for a few years now, so now i have a dependence on many neovim plugins (and related keybindings).

**Reply by JamesWidman** [reply]
> Did you open an issue and offer to help with testing and fixing?

I am going to try, but... 

1. I didn't set out to debug someone else's software. I was trying to work on my own non-neovim project when this broke!
2. It would be emotionally easier for me to debug someone else's software if I didn't get the impression that they're willing to just push out a change that breaks a big chunk of my keymaps.

**Reply by clason** [reply]
You know how open source software works, right? Did you even try to understand the change they made and why? (It fixed a _huge_ number of other input issues.)

**Reply by JamesWidman** [reply]
also: for what it's worth, i would love to be able to support GUI/neovim developers in a way that doesn't derail me. 

Like... I'm happy to pay for a software license or something! That way they can focus on making a good GUI, and I can focus on my stuff.

**Reply by JamesWidman** [reply]
> You know how open source software works, right?

i mean, right now, it doesn't.

> Did you even try to understand the change they made and why? (It fixed a huge number of other input issues.)

When I find out why it's not even launching, i will try to look into this.

**Reply by clason** [reply]
Again with the personal definition of "working" ;)

**Reply by JamesWidman** [reply]
> Again with the personal definition of "working" ;)

it doesn't reach main().

**Reply by JamesWidman** [reply]
> it doesn't reach main().

update: it turns out I'd left some debug-printing commands in a zsh init file, and that was messing with Neovide's attempt to detect a valid version of nvim, which it does by invoking `$SHELL -lc '/path/to/nvim -v'` (instead of just `/path/to/nvim -v`) and looking at the output, which neovide expects will begin with "NVIM v".

(neovide 0.10.4 (which is what I'm currently using so that keymaps work), was silent about this, but the current version of neovide gives a helpful error message. I didn't expect it would invoke nvim indirectly through a shell.  But I guess you want that if the nvim session is starting because someone double-clicked on the app—otherwise things like `$PATH` won't be set properly.  Anyway, since the current version of neovide gives a helpful error, I guess there's nothing to report to the neovide project.)

Next I'll try looking at the input stuff.

**Comment by JamesWidman** [comment]
I also think it's funny that, if there had been one separate & official GUI project for each of linux, mac, and windows, with no attempt to share code between these projects, and with each focused on producing a good experience for *one* operating system, there would have been much, much less duplication of effort compared to what we have now (with several projects using "cross-platform" APIs, whose main selling point is that they were supposed to help us *avoid* a duplication of effort).

**Reply by clason** [reply]
This "us" here is load-bearing...

**Comment by JamesWidman** [comment]
> So you have a very narrow definition of "working".

given that this is in the context of keymaps breaking, I think it's not that narrow. 

Please reconsider, if only on this one point.

**Reply by clason** [reply]
Nothing for us to reconsider -- GUIs are external projects, and from our point of view, this is still the right decision.

**Comment by bfredl** [comment]
> I was using neovide until today. A month ago, they broke <command-shift- bindings, [which is a known issue in version 0.11](https://github.com/neovide/neovide/releases) (see Breaking Changes)

So neovide had a regression. whoops. Now what makes you think this could never happen if the code was maintained in nvim core? We like to avoid it of course, but the master branch of neovim do have regressions from time to time.  if people who do focus on GUI:s cannot figure out command-shift-option-X whatever on OS X (at least not at the first try), no reason to assume that neovim core devs somehow will.

Also, there is nothing wrong with considering GTK4 as a platform. it is not for everyone, but that's the point. As a user, I'd like input methods to work, preferably regardless if my gnome-shell runs Xorg or wayland. GTK4 thus is a very welcome alternative for a user like me.

**Reply by JamesWidman** [reply]
[reposting because i messed up the reply-threading last time---oops]

> So neovide had a regression. whoops. Now what makes you think this could never happen if the code was maintained in nvim core?

>  if people who do focus on GUI:s cannot figure out command-shift-option-X whatever on OS X (at least not at the first try), no reason to assume that neovim core devs somehow will.

so... first of all, I don't expect, and did not mean to ask, for curent neovim core devs to get into GUI development.

By analogy, think of the LLVM project: if you look at [`clang/CodeOwners.rst`](https://github.com/llvm/llvm-project/blob/main/clang/CodeOwners.rst#clang-llvm-ir-generation) and compare it to [`lldb/CodeOwners.rst`](https://github.com/llvm/llvm-project/blob/main/lldb/CodeOwners.rst), you'll find that the person who owns the LLVM-IR-generating part of clang doesn't also own anything in the LLDB project (at least, not at the time that I'm writing this).

Similarly, if neovim were to adopt one or more GUI projects as the "official" GUI(s) of neovim (just like LLDB is the official debugger of LLVM), then people who own core parts of nvim would *NOT* also be responsible for GUI code (hopefully/probably).

Secondly: I recall that a neovide maintainer mentioned something to the effect that the neovide project doesn't currently have any maintainers who regularly use macOS.  Now, if there were an official neovim GUI project for macOS (which in theory could be neovide), then I bet that project would get more attention from prospective mac-GUI coders than alternative GUI projects.  And I think that would be a good thing, because it could solve the personnel problem reported by the neovide maintainer. It would improve the odds that the GUI for `$platform` will actually in a good state when it's time for it to be released.

**Reply by clason** [reply]
You would probably lose that bet. I use macOS and am a (not very active) maintainer of neovide. 

I cannot stress enough that your basic assumption underlying these "proposals" is simply false and not more than wishful thinking. The only thing that will ensure your issues are fixed is taking an active role in fixing them (as much as you are able; the more, the better your chances).

**Reply by JamesWidman** [reply]
> your basic assumption underlying these "proposals" is simply false

sorry; which assumption is that?

**Reply by clason** [reply]
That having a single "blessed" project leads to more development, rather than less.

**Reply by JamesWidman** [reply]
> That having a single "blessed" project leads to more development, rather than less.

I have to admit that I could be wrong about this!  Have you seen that happen (i.e. a reduction in development/interest after a project is blessed)?

**Reply by clason** [reply]
We've seen the converse happen, multiple times.

**Comment by bfredl** [comment]
Using gdb as the debugger for LLVM compiled code works just fine, and is on many platforms a better experience than using lldb for LLVM code. This works because both GNU and LLVM uses a shared format for debug info (DWARF). This better experience would **not** have been possible if lldb was mandated as the _only_ supported debugger for LLVM code.

It sounds like you wish there was simply more developers working on GUI tooling on mac OS X. A fair wish, but this would not magically happen sooner because GUI effort was organized as part of the neovim repo as opposed to, say, neovide.

**Reply by JamesWidman** [reply]
> This better experience would not have been possible if lldb was mandated as the only supported debugger for LLVM code.

I'm not suggesting that any GUI for neovim on a given platform should be the *only* GUI; just that it would help if one of them were the popular default (because it would focus more effort on one project rather than scattering that effort across multiple projects that serve the same platform).

**Reply by clason** [reply]
Yes, and I keep telling you, that's simply not how it works. Especially with GUIs, which nearly always are started because someone is either not happy with the fundamental design decisions of the existing ones ("too much like a TUI", "I just want a native TUI without terminal bloat", "I hate electron", ...) or just wants to create their own from scratch ("it can't be that hard! Right? ... right?!")

**Reply by JamesWidman** [reply]
I see your point...

I mean, part of me hopes that the ones who start their own projects for those reasons are the outliers among neovim-GUI coders rather than the majority...

But then, vim is an editor for people who become very particular about how their editor should behave, so I won't get those hopes up.

**Comment by zeertzjq** [comment]
Back to the original comment, you don't need a GUI to use all four modifiers. It should work when using Nvim TUI (nightly or upcoming v0.9.2) in kitty (not sure about other terminals).

**Reply by JamesWidman** [reply]
I was not aware! I will give kitty a try; thank you!
